File: packages/wasmvm/wasmlib/src/wasmtypes/scagentid.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub const SC_AGENT_ID_NIL: u8 = 0;
pub const SC_AGENT_ID_ADDRESS: u8 = 1;
pub const SC_AGENT_ID_CONTRACT: u8 = 2;
pub const SC_AGENT_ID_ETHEREUM: u8 = 3;
const NIL_AGENT_ID_STRING: &str = "-";

#[derive(PartialEq, Clone)]
pub struct ScAgentID {
    kind: u8,
    address: ScAddress,
    hname: ScHname,
    eth: ScAddress,
}

impl ScAgentID {
    pub fn new(address: &ScAddress, hname: ScHname) -> ScAgentID {
        ScAgentID {
            kind: SC_AGENT_ID_CONTRACT,
            address: address.clone(),
            hname: hname,
            eth: ScAddress {
                id: [0; SC_ADDRESS_LENGTH],
            },
        }
    }

    pub fn for_ethereum(chain: &ScAddress, eth_address: &ScAddress) -> ScAgentID {
        if chain.id[0] != SC_ADDRESS_ALIAS {
            panic("invalid eth AgentID: chain address");
        }
        if eth_address.id[0] != SC_ADDRESS_ETH {
            panic("invalid eth AgentID: eth address");
        }
        ScAgentID {
            kind: SC_AGENT_ID_ETHEREUM,
            address: chain.clone(),
            hname: ScHname(0),
            eth: eth_address.clone(),
        }
    }

    pub fn from_address(address: &ScAddress) -> ScAgentID {
        let kind;
        match address.id[0] {
            SC_ADDRESS_ALIAS => {
                kind = SC_AGENT_ID_CONTRACT;
            }
            SC_ADDRESS_ETH => {
                panic("invalid eth AgentID: need chain address");
                kind = 0;
            }
            _ => {
                kind = SC_AGENT_ID_ADDRESS;
            }
        }

        ScAgentID {
            kind: kind,
            address: address.clone(),
            hname: ScHname(0),
            eth: ScAddress {
                id: [0; SC_ADDRESS_LENGTH],
            },
        }
    }

    pub fn address(&self) -> ScAddress {
        self.address.clone()
    }

    pub fn eth_address(&self) -> ScAddress {
        self.eth.clone()
    }

    pub fn hname(&self) -> ScHname {
        self.hname
    }

    pub fn is_address(&self) -> bool {
        self.kind == SC_AGENT_ID_ADDRESS
    }

    pub fn is_contracts(&self) -> bool {
        self.kind == SC_AGENT_ID_CONTRACT
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        agent_id_to_bytes(self)
    }

    pub fn to_string(&self) -> String {
        agent_id_to_string(self)
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn agent_id_decode(dec: &mut WasmDecoder) -> ScAgentID {
    agent_id_from_bytes(&dec.bytes())
}

pub fn agent_id_encode(enc: &mut WasmEncoder, value: &ScAgentID) {
    enc.bytes(&agent_id_to_bytes(value));
}

pub fn agent_id_from_bytes(buf: &[u8]) -> ScAgentID {
    let len = buf.len();
    if len == 0 {
        return ScAgentID {
            kind: SC_AGENT_ID_NIL,
            address: address_from_bytes(buf),
            hname: ScHname(0),
            eth: ScAddress {
                id: [0; SC_ADDRESS_LENGTH],
            },
        };
    }
    let len = len - 1;
    match buf[0] {
        SC_AGENT_ID_ADDRESS => {
            let buf: &[u8] = &buf[1..];
            if len != SC_LENGTH_ALIAS && len != SC_LENGTH_ED25519 {
                panic("invalid AgentID length: address agentID");
            }
            return ScAgentID::from_address(&address_from_bytes(&buf));
        }
        SC_AGENT_ID_CONTRACT => {
            let buf: &[u8] = &buf[1..];
            if len != SC_CHAIN_ID_LENGTH + SC_HNAME_LENGTH {
                panic("invalid AgentID length: contract agentID");
            }
            let chain_id = chain_id_from_bytes(&buf[..SC_CHAIN_ID_LENGTH]);
            let hname = hname_from_bytes(&buf[SC_CHAIN_ID_LENGTH..]);
            return ScAgentID::new(&chain_id.address(), hname);
        }
        SC_AGENT_ID_ETHEREUM => {
            let buf: &[u8] = &buf[1..];
            if len != SC_CHAIN_ID_LENGTH + SC_LENGTH_ETH {
                panic("invalid AgentID length: eth agentID");
            }
            let chain_id = chain_id_from_bytes(&buf[..SC_CHAIN_ID_LENGTH]);
            let eth_address = address_from_bytes(&buf[SC_CHAIN_ID_LENGTH..]);
            return ScAgentID::for_ethereum(&chain_id.address(), &eth_address);
        }
        SC_AGENT_ID_NIL => {}
        _ => panic("AgentIDFromBytes: invalid AgentID type"),
    }
    ScAgentID {
        kind: SC_AGENT_ID_NIL,
        address: address_from_bytes(&[]),
        hname: ScHname(0),
        eth: ScAddress {
            id: [0; SC_ADDRESS_LENGTH],
        },
    }
}

pub fn agent_id_to_bytes(value: &ScAgentID) -> Vec<u8> {
    let mut buf: Vec<u8> = Vec::new();
    buf.push(value.kind);
    match value.kind {
        SC_AGENT_ID_ADDRESS => {
            buf.extend_from_slice(&address_to_bytes(&value.address));
        }
        SC_AGENT_ID_CONTRACT => {
            buf.extend_from_slice(&address_to_bytes(&value.address)[1..]);
            buf.extend_from_slice(&hname_to_bytes(value.hname));
        }
        SC_AGENT_ID_ETHEREUM => {
            buf.extend_from_slice(&address_to_bytes(&value.address)[1..]);
            buf.extend_from_slice(&address_to_bytes(&value.eth));
        }
        SC_AGENT_ID_NIL => (),
        _ => panic("AgentIDToBytes: invalid AgentID type"),
    }
    buf
}

pub fn agent_id_from_string(value: &str) -> ScAgentID {
    if value.eq(NIL_AGENT_ID_STRING) {
        return agent_id_from_bytes(&[]);
    }

    let parts: Vec<&str> = value.split("@").collect();
    return match parts.len() {
        1 => ScAgentID::from_address(&address_from_string(&parts[0])),
        2 => {
            if !value.starts_with("0x") {
                return ScAgentID::new(
                    &address_from_string(&parts[1]),
                    hname_from_string(&parts[0]),
                );
            }
            return ScAgentID::for_ethereum(
                &address_from_string(&parts[1]),
                &address_from_string(&parts[0]),
            );
        },
        _ => {
            panic("invalid AgentID string");
            agent_id_from_bytes(&[])
        }
    };
}

pub fn agent_id_to_string(value: &ScAgentID) -> String {
    match value.kind {
        SC_AGENT_ID_ADDRESS => {
            return value.address().to_string();
        }
        SC_AGENT_ID_CONTRACT => {
            return value.hname().to_string() + "@" + &value.address().to_string();
        }
        SC_AGENT_ID_ETHEREUM => {
            return value.eth_address().to_string() + "@" + &value.address().to_string();
        }
        SC_AGENT_ID_NIL => {
            return NIL_AGENT_ID_STRING.to_string();
        }
        _ => panic("AgentIDToString: invalid AgentID type"),
    }
    "".to_string()
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableAgentID {
    proxy: Proxy,
}

impl ScImmutableAgentID {
    pub fn new(proxy: Proxy) -> ScImmutableAgentID {
        ScImmutableAgentID { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        agent_id_to_string(&self.value())
    }

    pub fn value(&self) -> ScAgentID {
        agent_id_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable ScAgentID in host container
pub struct ScMutableAgentID {
    proxy: Proxy,
}

impl ScMutableAgentID {
    pub fn new(proxy: Proxy) -> ScMutableAgentID {
        ScMutableAgentID { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: &ScAgentID) {
        self.proxy.set(&agent_id_to_bytes(&value));
    }

    pub fn to_string(&self) -> String {
        agent_id_to_string(&self.value())
    }

    pub fn value(&self) -> ScAgentID {
        agent_id_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/scagentid.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from '../sandbox';
import {
    addressFromBytes,
    addressFromString,
    addressToBytes,
    addressToString,
    ScAddress,
    ScAddressAlias,
    ScAddressEth,
    ScAddressEthLength,
    ScLengthAlias,
    ScLengthEd25519
} from './scaddress';
import {chainIDFromBytes, ScChainIDLength} from './scchainid';
import {hnameFromBytes, hnameFromString, hnameToBytes, hnameToString, ScHname, ScHnameLength} from './schname';
import {concat, WasmDecoder, WasmEncoder} from './codec';
import {Proxy} from './proxy';

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export const ScAgentIDNil: u8 = 0;
export const ScAgentIDAddress: u8 = 1;
export const ScAgentIDContract: u8 = 2;
export const ScAgentIDEthereum: u8 = 3;
const nilAgentIDString = '-';

export class ScAgentID {
    kind: u8;
    _address: ScAddress;
    _hname: ScHname;

    constructor(address: ScAddress, hname: ScHname) {
        this.kind = ScAgentIDContract;
        this._address = address;
        this._hname = hname;
    }

    public static fromAddress(address: ScAddress): ScAgentID {
        const agentID = new ScAgentID(address, new ScHname(0));
        switch (address.id[0]) {
            case ScAddressAlias: {
                break;
            }
            case ScAddressEth: {
                agentID.kind = ScAgentIDEthereum;
                break;
            }
            default: {
                agentID.kind = ScAgentIDAddress;
                break;
            }
        }
        return agentID;
    }

    public equals(other: ScAgentID): bool {
        return this._address.equals(other._address) &&
            this._hname.equals(other._hname);
    }

    public address(): ScAddress {
        return this._address;
    }

    public hname(): ScHname {
        return this._hname;
    }

    public isAddress(): bool {
        return this.kind == ScAgentIDAddress;
    }

    public isContract(): bool {
        return this.kind == ScAgentIDContract;
    }

    // convert to byte array representation
    public toBytes(): Uint8Array {
        return agentIDToBytes(this);
    }

    // human-readable string representation
    public toString(): string {
        return agentIDToString(this);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function agentIDDecode(dec: WasmDecoder): ScAgentID {
    return agentIDFromBytes(dec.bytes());
}

export function agentIDEncode(enc: WasmEncoder, value: ScAgentID): void {
    enc.bytes(agentIDToBytes(value));
}

export function agentIDFromBytes(buf: Uint8Array | null): ScAgentID {
    if (buf === null || buf.length == 0) {
        const agentID = ScAgentID.fromAddress(addressFromBytes(null));
        agentID.kind = ScAgentIDNil;
        return agentID;
    }
    switch (buf[0]) {
        case ScAgentIDAddress: {
            buf = buf.subarray(1);
            if (buf.length != ScLengthAlias && buf.length != ScLengthEd25519) {
                panic('invalid AgentID length: address agentID');
            }
            return ScAgentID.fromAddress(addressFromBytes(buf));
        }
        case ScAgentIDContract: {
            buf = buf.subarray(1);
            if (buf.length != ScChainIDLength + ScHnameLength) {
                panic('invalid AgentID length: contract agentID');
            }
            const chainID = chainIDFromBytes(buf.subarray(0, ScChainIDLength));
            const hname = hnameFromBytes(buf.subarray(ScChainIDLength));
            return new ScAgentID(chainID.address(), hname);
        }
        case ScAgentIDEthereum:
            buf = buf.subarray(1);
            if (buf.length != ScAddressEthLength) {
                panic('invalid AgentID length: Eth agentID');
            }
            return ScAgentID.fromAddress(addressFromBytes(buf));
        case ScAgentIDNil:
            break;
        default: {
            panic('AgentIDFromBytes: invalid AgentID type');
            break;
        }
    }
    return agentIDFromBytes(null);
}

export function agentIDToBytes(value: ScAgentID): Uint8Array {
    let buf = new Uint8Array(1);
    buf[0] = value.kind;
    switch (value.kind) {
        case ScAgentIDAddress:
            return concat(buf, addressToBytes(value._address));
        case ScAgentIDContract: {
            buf = addressToBytes(value._address);
            buf[0] = value.kind;
            return concat(buf, hnameToBytes(value._hname));
        }
        case ScAgentIDEthereum:
            return concat(buf, addressToBytes(value._address));
        case ScAgentIDNil:
            return buf;
        default: {
            panic('AgentIDToBytes: invalid AgentID type');
            break;
        }
    }
    return buf;
}

export function agentIDFromString(value: string): ScAgentID {
    if (value == nilAgentIDString) {
        return agentIDFromBytes(null);
    }

    const parts = value.split('@');
    switch (parts.length) {
        case 1:
            return ScAgentID.fromAddress(addressFromString(parts[0]));
        case 2:
            return new ScAgentID(addressFromString(parts[1]), hnameFromString(parts[0]));
        default:
            panic('invalid AgentID string');
            return agentIDFromBytes(null);
    }
}

export function agentIDToString(value: ScAgentID): string {
    switch (value.kind) {
        case ScAgentIDAddress:
            return addressToString(value.address());
        case ScAgentIDContract: {
            return hnameToString(value.hname()) + '@' + addressToString(value.address());
        }
        case ScAgentIDEthereum:
            return addressToString(value.address());
        case ScAgentIDNil:
            return nilAgentIDString;
        default: {
            panic('AgentIDToString: invalid AgentID type');
            return '';
        }
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableAgentID {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return agentIDToString(this.value());
    }

    value(): ScAgentID {
        return agentIDFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableAgentID extends ScImmutableAgentID {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: ScAgentID): void {
        this.proxy.set(agentIDToBytes(value));
    }
}
